### PR TITLE
Connect mocked graphql structure to real data

### PR DESCRIPTION
**Pull request by jvaclavik**
Part of #123 

### Example

**Query**
```
query RootQuery($input: ItinerariesSearchInput!){
  searchItineraries (input: $input){
    edges{
      node{
        id
        type
        price {
          amount
          currency
        }
        origin {
          id
          locationId
          name
          timezone
          country {
            name
            code
          }
        }
        destination {
          id
          locationId
          name
          timezone
          country {
            name
            code
          }
        }
        startTime {
          local
          utc
        }
        endTime {
          local
          utc
        }
        sectors{
          id
          departureTime{
            local
            utc
          }
          arrivalTime{
            local
            utc
          }
          destination{
            name
            locationId
          }
          origin{
            name
            locationId
          }
          duration
          segments{
            arrivalTime {
              utc
            }
            departureTime{
              utc
            }
            destination{
              name
              locationId
            }
            duration
            origin{
              name
              locationId
            }
            transporter {
              name
            }
            vehicle{
              type
              uniqueNo
            }
          }
        }
      }
    }
  } 
}
```

**Query variables**
```
{
  "input": {
    "travelFrom":"OSL",
    "travelTo": "PRG",
    "dateFrom": "2019-03-10",
    "dateTo": "2019-03-10",
    "returnDateFrom": "2019-05-13",
    "returnDateTo": "2019-05-13"
  }
}
```

### CODE REVIEW

File: apps/graphql/src/dataloaders/itinerariesHelpers.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  flagURL: 'flagUrl',
+});
+
+export const getLocation = (
+  locationId?: string,
+  name?: string,
+  countryName?: string,
+  countryCode?: string,
+) => ({
+  id: '',
```
Comment (FilipMessa) [inline apps/graphql/src/dataloaders/itinerariesHelpers.js]: I have been thinking about and maby we can avoid this `locationId` and instead of this use just an `id`.
It will be encoded by GlobalID and if we need it we can decode it any time.

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 2]
```diff
@@ -58,40 +65,108 @@ const fetchItineraries = async (
...
+            connections: [],
+            departureTime: getDate(
+              firstSegment.local_departure,
+              firstSegment.utc_departure,
+            ),
+            duration: DateFNS.differenceInMinutes(
+              lastSegment.utc_arrival,
+              firstSegment.utc_departure,
+            ),
+            id: 'asdfd',
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: I don't think so, because this `id` is specific for one Sector. And `itinerary_id` is specific for one `Itinerary`

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 3]
```diff
@@ -58,40 +65,108 @@ const fetchItineraries = async (
...
+                departureTime: getDate(
+                  segment.local_departure,
+                  segment.utc_departure,
+                ),
+                destination: getLocation(segment.flyTo, segment.cityTo),
+                duration: DateFNS.differenceInMinutes(
+                  segment.utc_arrival,
+                  segment.utc_departure,
+                ),
+                id: segment.id ?? '',
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: It shouldn't be there. I will delete it

File: apps/graphql/src/dataloaders/itinerariesHelpers.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  ProviderTypeType,
+} from './ItinerariesloaderTypes';
+
+export const getVehicle = (type: ?ProviderTypeType, uniqueNo: string) => ({
+  type: type ?? null,
+  uniqueNo: uniqueNo ?? null,
+});
+
+export const getProvider = (provider: ?string) => ({
+  id: 'ads',
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/itinerariesHelpers.js]: I didn't find anything like that. 

Btw one more thing – `provider` is probably used on Tequila as a data provider a not as a travel provider, because I found there following row: 
`"check_price_provider": "AMADEUS_MRP",`

[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/itinerariesHelpers.js]: it's `transporter` now

File: apps/graphql/src/dataloaders/ItinerariesloaderTypes.js [hunk 5]
```diff
@@ -30,68 +30,71 @@ export type RouteItemType = {|
...
-  +local: Date,
-  +utc: Date,
+  +local: ?Date,
+  +utc: ?Date,
 |};
-export type ProviderType = {|
-  +id: string,
-  +name: string,
+
+export type VehicleTypeType = 'bus' | 'train' | 'aircraft';
```
Comment (FilipMessa) [inline apps/graphql/src/dataloaders/ItinerariesloaderTypes.js]: Same as below, we can relay on API response.

File: apps/graphql/src/dataloaders/itinerariesHelpers.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+export const getDate = (local: ?Date, utc: ?Date) => ({
+  local: local ?? null,
+  utc: utc ?? null,
+});
+
+export const getCountry = (name: ?string, code: ?string) => ({
+  id: code ?? '',
+  name: name ?? '',
+  code: code ?? '',
```
Comment (FilipMessa) [inline apps/graphql/src/dataloaders/itinerariesHelpers.js]: Shouldn't we discussed that we want to return `null` instead of empty string? 
Same with `getLocation`

File: apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+describe('getVehicle', () => {
+  const type = 'aircraft';
+  const uniqueNo = '123456';
+  it('returns proper structure of the Vehicle object', () => {
+    expect(getVehicle(type, uniqueNo)).toMatchObject({
+      type: type,
+      uniqueNo: uniqueNo,
+    });
+  });
+});
```
Comment (FilipMessa) [inline apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js]: What about also add some test for edge cases? like missing data etc...

**Comment by FilipMessa** [comment]
Regarding the `getDate` and all this `getWhatever` I'm still not sure about naming `sanitizeDate` make more sense for me. 

What do you think guys? @tbergq @RobinCsl @JoseDuSV

**Comment by tbergq** [comment]
For me, the name get means you are getting the data from somewhere else. It could be from an object or database of whatever. If this refers to just map an object we get from the restAPI to our structure, then definitely `sanitize` or even `map` is more correct I would say.

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 8]
```diff
@@ -58,40 +65,107 @@ const fetchItineraries = async (
 const sanitizeItineraries = (response: ApiResponseType): ItinerariesType[] => {
   const itineraries = response.data;
 
-  const newItinerariesStructure = itineraryMock;
+  return itineraries.map(itinerary => {
+    return {
+      // new structure
```
Comment (tbergq) [inline apps/graphql/src/dataloaders/Itinerariesloader.js]: I think this comment is abundant

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 9]
```diff
@@ -58,40 +65,107 @@ const fetchItineraries = async (
...
+      type: getItineraryType(itinerary.routes),
+      startTime: getDate(itinerary.local_departure, itinerary.utc_departure),
+      endTime: getDate(itinerary.local_arrival, itinerary.utc_arrival),
+      destination: getLocation(
+        itinerary.flyTo,
+        itinerary.cityTo,
+        itinerary.countryFrom && itinerary.countryTo.name,
+        itinerary.countryFrom && itinerary.countryTo.code,
```
[the remark this answers is not in the document]
Comment (tbergq) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: `itinerary.countryTo?.name` & `itinerary.countryTo?.code`

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 10]
```diff
@@ -58,40 +65,107 @@ const fetchItineraries = async (
...
+      origin: getLocation(
+        itinerary.flyFrom,
+        itinerary.cityFrom,
+        itinerary.countryFrom && itinerary.countryFrom.name,
+        itinerary.countryFrom && itinerary.countryFrom.code,
+      ),
+      sectors:
+        itinerary.route &&
+        itinerary.routes &&
+        getSectors(itinerary.route, itinerary.routes).map(sector => {
```
Comment (tbergq) [inline apps/graphql/src/dataloaders/Itinerariesloader.js]: Why not just pass itinerary.route and itinerary.routes and let the getSectors method handle if it is undefined?

File: apps/graphql/src/types/output/Itinerary.js [hunk 11]
```diff
@@ -77,10 +77,9 @@ const VehicleType = new GraphQLObjectType({
     uniqueNo: { type: GraphQLString },
   },
 });
-const ProviderType = new GraphQLObjectType({
-  name: 'ProviderType',
+const TransporterType = new GraphQLObjectType({
```
Comment (tbergq) [inline apps/graphql/src/types/output/Itinerary.js]: Could we stick to one `GraphQLObjectType` pr file?

File: apps/graphql/src/dataloaders/itinerariesHelpers.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  if (routes.length === 1) {
+    return 'oneway';
+  }
+  if (routes.length === 2) {
+    return 'return';
+  }
+  return null;
+};
+
+export const getSectors = (
```
Comment (tbergq) [inline apps/graphql/src/dataloaders/itinerariesHelpers.js]: Also I fail to understand what this function does.

And why do we have to call map on the function afterwards?
`getSectors(itinerary.route, itinerary.routes).map`

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 13]
```diff
@@ -58,40 +65,107 @@ const fetchItineraries = async (
...
+          const firstSegment = sector[0];
+          const lastSegment = sector[sector.length - 1];
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: I think so. And if not it would be bug in API, because it wouldn't make sense.

File: apps/graphql/src/dataloaders/ItinerariesloaderTypes.js [hunk 14]
```diff
@@ -30,68 +30,71 @@ export type RouteItemType = {|
...
 export type newItinerariesStructureType = {|
   id: string,
-  +type: 'oneway' | 'return' | 'multicity',
-  +price: PriceType,
-  +origin: LocationType,
-  +destination: LocationType,
-  +startTime: DateType,
-  +endTime: DateType,
-  +sectors: Array<SectorType>,
+  +type: ?'oneway' | 'return' | 'multicity',
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/ItinerariesloaderTypes.js]: Did you mean: "we can rely on that that we will NOT receive one of this type."? So should it be string?

File: apps/graphql/src/dataloaders/Itinerariesloader.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (FilipMessa) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: But we always can sort that data by `utc_departure`, right?

File: apps/graphql/src/dataloaders/ItinerariesloaderTypes.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (FilipMessa) [reply apps/graphql/src/dataloaders/ItinerariesloaderTypes.js]: Yeas I mean we are not owners of the Tequila API so we can be sure that we always get back one of these `'oneway' | 'return' | 'multicity'` values.
So string makes more sense here.

File: apps/graphql/src/types/output/Itinerary.js (hunk 11 above)
Comment (FilipMessa) [reply apps/graphql/src/types/output/Itinerary.js]: 👍 good point will be better for the future.

File: apps/graphql/src/dataloaders/itinerariesHelpers.js (hunk 12 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/itinerariesHelpers.js]: This function should get list of the flights and some map which defines how the flight should be divided into sectors (see what is sector here: https://kiwi.wiki/kiwi/handbook/#/comms/glossary?id=itinerary-parts-examples):

We need it for example to calculate duration in your final destination:

screenshot 2019-01-11 at 16 10 34


Example:

**List of the flights:**
```js
[
  {
    flyFrom: 'OSL',
    flyTo: 'BCN',
    ...
  },
  {
    flyFrom: 'BCN',
    flyTo: 'PRG',
    ...
  },
  {
    flyFrom: 'PRG',
    flyTo: 'OSL',
    ...
  },
]
```

**Map:**
```
[
  ['OSL', 'PRG'], 
  ['PRG', 'OSL']
]
```

**Output from function:**

```js
[
  [
    {
      flyFrom: 'OSL',
      flyTo: 'BCN',
      ...
    },
    {
      flyFrom: 'BCN',
      flyTo: 'PRG',
      ...
    },
  ],
  [
    {
      flyFrom: 'PRG',
      flyTo: 'OSL',
      ...
    },
  ],
]
```

File: apps/graphql/src/dataloaders/Itinerariesloader.js [hunk 15]
```diff
@@ -58,40 +65,107 @@ const fetchItineraries = async (
...
+          return {
+            destination: getLocation(lastSegment.flyTo, lastSegment.cityTo),
+            origin: getLocation(firstSegment.flyFrom, firstSegment.cityFrom),
+            arrivalTime: getDate(
+              lastSegment.local_arrival,
+              lastSegment.utc_arrival,
+            ),
+            connections: [],
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: No, deleted

File: apps/graphql/src/dataloaders/Itinerariesloader.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: done

File: apps/graphql/src/dataloaders/ItinerariesloaderTypes.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/ItinerariesloaderTypes.js]: Done

File: apps/graphql/src/dataloaders/ItinerariesloaderTypes.js (hunk 5 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/ItinerariesloaderTypes.js]: Done

File: apps/graphql/src/dataloaders/itinerariesHelpers.js (hunk 6 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/itinerariesHelpers.js]: Done

File: apps/graphql/src/dataloaders/Itinerariesloader.js (hunk 8 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: Removed

File: apps/graphql/src/dataloaders/Itinerariesloader.js (hunk 10 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: Done

File: apps/graphql/src/dataloaders/ItinerariesloaderTypes.js [hunk 16]
```diff
@@ -112,25 +115,38 @@ export type ApiRouteItemType = {|
...
-  +utc_arrival?: ?Date,
   +local_departure?: Date,
+  +utc_arrival?: ?Date,
   +utc_departure?: Date,
+  +vehicle_type?: VehicleTypeType,
+|};
+
+export type ApiCountryType = {|
+  code: string,
+  name: string,
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/ItinerariesloaderTypes.js]: Fixed

File: apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js (hunk 7 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js]: Done

**Comment by jvaclavik** [comment]
> For me, the name get means you are getting the data from somewhere else. It could be from an object or database of whatever. If this refers to just map an object we get from the restAPI to our structure, then definitely `sanitize` or even `map` is more correct I would say.

Done

File: apps/graphql/src/dataloaders/Itinerariesloader.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Itinerariesloader.js]: OK, done

File: apps/core/scenes/results/ResultsListItem.js [hunk 17]
```diff
@@ -33,7 +33,9 @@ class ResultListItem extends React.Component<Props> {
   getSanitizedRouteItem = (routeItem: RouteItem) => {
     const timeSimpleFormat = 'H:mm';
 
-    const localArrival = this.parseDate(routeItem?.arrival?.localTime);
+    const localArrival =
+      routeItem?.arrival?.localTime &&
+      this.parseDate(routeItem?.arrival?.localTime);
```
Comment (FilipMessa) [inline apps/core/scenes/results/ResultsListItem.js]: Why this change?

File: apps/graphql/src/dataloaders/Locationsloader.js [hunk 18]
```diff
@@ -5,28 +5,24 @@ import qs from 'querystring';
...
 
 function sanitizeLocations(locations: $PropertyType<ApiResponse, 'locations'>) {
   return locations.map(location => ({
     id: location.id,
+    locationId: location.id,
     name: location.name,
     slug: location.slug,
+    timezone: location.timezone,
+    country: {
+      id: location.city?.country?.id ?? '',
```
Comment (FilipMessa) [inline apps/graphql/src/dataloaders/Locationsloader.js]:
```suggestion
      id: location.city?.country?.id,
```

File: apps/graphql/src/dataloaders/itinerariesHelpers.js [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+): Array<ApiRouteItem> => {
+  return segments.slice(0).sort((segmentA, segmentB) => {
+    const a = segmentA.utc_departure && new Date(segmentA.utc_departure);
+    const b = segmentB.utc_departure && new Date(segmentB.utc_departure);
+    if (!a || !b) return 0;
+    return a < b ? -1 : a > b ? 1 : 0;
+  });
+};
+
+export const mapSectors = (
```
Comment (FilipMessa) [inline apps/graphql/src/dataloaders/itinerariesHelpers.js]: Will be possible to add some explanation comment to this function because it is really hard to understand what's going on there

File: apps/graphql/src/mocks/itinerary.js [hunk 20]
```diff
@@ -1,19 +1,19 @@
 // @flow
 
-const getCountry = () => ({
+const mapLocationArea = () => ({
```
Comment (FilipMessa) [inline apps/graphql/src/mocks/itinerary.js]: should not just object be enough?

File: apps/graphql/src/types/output/LocationArea.js [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+
+export default new GraphQLObjectType({
+  name: 'LocationArea',
+  fields: {
+    id: GlobalID(({ id }) => id),
+    locationId: { type: GraphQLString },
+    name: { type: GraphQLString },
+    slug: { type: GraphQLString },
+    code: { type: GraphQLString },
+  },
```
Comment (FilipMessa) [inline apps/graphql/src/types/output/LocationArea.js]: what about `'flagURL': { type: GraphQLString }` ?

File: apps/graphql/src/dataloaders/Locationsloader.js (hunk 18 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/Locationsloader.js]: screenshot 2019-01-31 at 14 32 37

File: apps/core/scenes/results/ResultsListItem.js (hunk 17 above)
Comment (jvaclavik) [reply apps/core/scenes/results/ResultsListItem.js]: It wasn't necessary, so I it's deleted now

File: apps/graphql/src/types/output/Segment.js [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+import Location from './Location';
+
+export default new GraphQLObjectType({
+  name: 'Segment',
+  fields: {
+    arrivalTime: { type: DateType },
+    departureTime: { type: DateType },
+    destination: { type: Location },
+    duration: { type: GraphQLInt },
+    id: GlobalID(({ id }) => id),
```
Comment (FilipMessa) [inline apps/graphql/src/types/output/Segment.js]: can you move ID to the top, please?

File: apps/graphql/src/dataloaders/Locationsloader.js (hunk 18 above)
Comment (FilipMessa) [reply apps/graphql/src/dataloaders/Locationsloader.js]: So probably OK then I don't have better idea

File: apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+  twoWayRoutesMap,
+  unsortedRoute,
+  sortedRoute,
+} from '../__mocks__/itinerariesMock';
+
+describe('mapLocationArea', () => {
+  const name = 'Czech republic';
+  const code = 'CZ';
+  it('returns proper structure of the Location Area object', () => {
+    expect(mapLocationArea(code, code, name)).toMatchObject({
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js]: OK, done

File: apps/graphql/src/dataloaders/LocationsloaderTypes.js [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+  +locationId: ?string,
+  +name: ?string,
+  +timezone: ?string,
+  +country: ?LocationArea,
+  +slug: ?string,
+|};
+
+export type ApiResponse = {|
+  +locations: $ReadOnlyArray<{|
+    +id: string,
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/LocationsloaderTypes.js]: Yes, done

File: apps/graphql/src/dataloaders/LocationsloaderTypes.js [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+    +locationId: ?string,
+    +name: ?string,
+    +slug: ?string,
+    +timezone: ?string,
+    +city: ?{
+      +country: ?{
+        id: string,
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/LocationsloaderTypes.js]: Done

File: apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+      id: code,
+      locationId: code,
+      name: name,
+      code: code,
+      slug: null,
+    });
+  });
+  it('returns empty structure of the Location Area object', () => {
+    expect(mapLocationArea()).toMatchObject({
+      id: '',
```
[the remark this answers is not in the document]
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/__tests__/itinerariesHelpers.test.js]: OK, not anymore :)

File: apps/graphql/src/types/output/LocationArea.js (hunk 21 above)
Comment (FilipMessa) [reply apps/graphql/src/types/output/LocationArea.js]: `countryFlagURL` shouldn't be necessary  it shoul be inside `country` and country is `LocationArea` type so there you can find that flagURL

File: apps/graphql/src/types/output/Segment.js (hunk 22 above)
Comment (FilipMessa) [reply apps/graphql/src/types/output/Segment.js]: All other can be sort alphabetically is nice but the ID should be always as first.

File: apps/graphql/src/types/output/LocationArea.js (hunk 21 above)
Comment (jvaclavik) [reply apps/graphql/src/types/output/LocationArea.js]: It's more logical, that's true. So I changed it.

File: apps/graphql/src/dataloaders/itinerariesHelpers.js (hunk 19 above)
Comment (FilipMessa) [reply apps/graphql/src/dataloaders/itinerariesHelpers.js]: This is not what I mean 😃. 
Try to imagine that you will look at the `mapSectors` after a few mounts.
I really don't understand what's going on. 
Try to explain what happens in first map iteration on routesMaps, this is the most confusing part. thanks

**Comment by FilipMessa** [comment]
May I ask you to write down which fields in the schema are not already connected to API? Thanks, It should be for future issues.

**Comment by jvaclavik** [comment]
> May I ask you to write down which fields in the schema are not already connected to API? Thanks, It should be for future issues.

```
location.slug
location.timezone
mapLocationArea.slug
mapLocationArea.flagURL
```

File: apps/graphql/src/dataloaders/itinerariesHelpers.js (hunk 19 above)
Comment (jvaclavik) [reply apps/graphql/src/dataloaders/itinerariesHelpers.js]: Done
